### ISSUE TITLE
Replace nodename, nodeport in pg_dist_shard_placement with groupid

### ISSUE THREAD

**Issue by lithp** [citusdata/citus#1414]
**Proposal**

Drop `nodename` and `nodeport` from `pg_dist_shard_placement` and add a `groupid` column. Making the new schema:

```
                            Table "pg_dist_shard_placement"
   Column    |  Type   |                              Modifiers                               
-------------+---------+----------------------------------------------------------------------
 shardid     | bigint  | not null
 shardstate  | integer | not null
 shardlength | bigint  | not null
 placementid | bigint  | not null default nextval('pg_dist_shard_placement_placementid_seq'::regclass)
 groupid     | integer | not null
```

There is some work planned for 7.0 which this will enable:
- Integrate Citus with auto-HA
- (5079 from the planning sheet) cross-region replication, followers in a region for low latency reads

Also related is a ticket from Cloud:
- https://github.com/citusdata/citus/issues/947, specify full connection strings for talking to workers, not just node{name,port}
---

**Backwards Compatibility**

This is a breaking change, fitting of our policy of not making such changes until a big release :)

We could lessen the impact by instead creating a new table `pg_dist_placement` and replacing the existing `pg_dist_shard_placement` with a view, but that will be confusing for users wondering why there are two tables showing the same information, and will complicate full connection string support.

By removing two columns and adding a column, we're ensuring that scripts which use the nodename/nodeport columns fail entirely. They're unlikely to continue running and return incorrect results.

So, I think it'd be better to make a clean break and not attempt to make this backwards compatible.

**Comment by marcocitus** [comment]
`nodeid` should probably be `groupid` to indicate that a shard placement is stored on all nodes in a group of primary/standby/follower nodes.

**Comment by lithp** [comment]
Oops, that's embarrassing. Fixed.

**Comment by lithp** [comment]
**Motivation**
---

**Auto-HA**

For auto-HA we want to start routing requests to the new primary once the old primary has failed. Currently this looks like:

- Update all shard placements to point to the secondary once it's been promoted (currently means running `UPDATE pg_dist_shard_placement SET nodename, nodeport WHERE nodename = ...` which is kind of ugly and takes a lot of locks)
- Separately update `nodename` and `nodeport` in `pg_dist_node`.

Preferably, we would only have to update `pg_dist_node`.

Having a specific `nodename` and `nodeport` in `pg_dist_shard_placement` also prevents us from adding logic to query secondaries during a failover.

---

**Cross-Region**

Reliance on (nodename, nodeport) is also a problem for multi-region. If we have a coordinator-follower in the second region we're not able to run `UPDATE`. 

Currently you'd need to do either of:
- Use iptables to intercept connections from the coordinator-follower to the workers and reroute them to the worker-followers instead.
- Always use ('localhost', nodeport) and run pgbouncer, configured differently on the different regions

But it'd be nicer if we:
- Add a column to `pg_dist_node`, something like `regionid`
- Add all nodes to `pg_dist_node` and include which region they're in
- Set a `GUC` on the coordinators, telling them which region to send requests to.

This is made a lot easier by the proposed change to `pg_dist_shard_placement`. I suppose we could also add `regionid` to the shard placements but ick.

---

**Full connection strings**

It's possible to add a column to `pg_dist_node` with the proper connection string, or maybe create a `pg_dist_user (nodeid, rolename, connstr)`, but doing that without removing nodename and nodeport from pg_dist_shard_placement is just asking for support requests.

**Comment by lithp** [comment]
@samay-sharma, do you have an idea of how much this will impact our existing customers? Will the migration be particularly difficult for any clients?

**Comment by metdos** [comment]
@ozgune, @marcocitus this is ([Brian's comment](https://github.com/citusdata/citus/issues/1414#issuecomment-301789413)) like what we need for "Reads to secondary nodes: -- Low latency reads in a region & database followers". What else do we need to cross that task?

**Comment by samay-sharma** [comment]
@lithp : Hmm, this is a pretty big change.

One thing it'll break are the run_on_all_* functions. However, as they are a part of regression tests, I'm assuming we'll fix them.

Several of our customers do rely on xargs scripts to do things on the worker nodes and I remember recommending / suggesting it to a few others. I'm sure @marcocitus has also used / suggested it several times. Even searching on Citus users results 4 such results.

I know atleast 3-4 customers have used https://gist.github.com/samay-sharma/06852d9e7f7b08fe077a2c4e45eb3185 and 1 of them has also created their own fork of this which allows you to take in a lot of options.

The run_on_all_* functions do reduce the need to use those columns but there are still certain use cases where using them makes sense. And for the customers we on boarded before those tools existed, this was the way of running something on the shards.

So, in terms of impact, I think it is quite a few customers. Also, since this is customer scripts, it's not in our control. I also don't know if Cloud uses these columns (cc @lfittl) 

Also, if we have to make this change to proceed, I'd prefer that this is heavily emphasized in the docs, and put in the upgrade process notes for the 7.0 release.

**Comment by samay-sharma** [comment]
Also, a side note, we use that table in almost all hands on sessions to demo how Citus works under the hood. This change will add some friction to that demo, but that we can change.

**Comment by anarazel** [comment]
> Also, a side note, we use that table in almost all hands on sessions to demo how Citus works under the hood. This change will add some friction to that demo, but that we can change.

In the planning meeting about this we'd pretty much decided that we'd
replace pg_dist_shard_placement with a view, presenting the old data.
The real data would be in something like pg_dist_placement.   Then we
can deprecate pg_dist_shard_placement and drop it in a few years.

@samaysharma: That should address most of your concerns?

**Comment by lithp** [comment]
Thanks for the feedback @samay-sharma,

Agree that we'll want to stress this in the documentation no matter what we do, even if we deprecate `pg_dist_shard_placement` and turn it into a view we'll want to get rid of it sooner than later.

Do you have an idea of how hard it would be for our customers to migrate? The query you linked:

```sql
SELECT s.shardid, nodename, nodeport
FROM pg_dist_shard s
JOIN pg_dist_shard_placement p ON (s.shardid = p.shardid)
WHERE logicalrelid::regclass = '$TABLE_NAME'::regclass;
```

would likely turn into:

```sql
SELECT s.shardid, nodename, nodeport
FROM pg_dist_shard s
JOIN pg_dist_shard_placement p ON (s.shardid = p.shardid)
JOIN pg_dist_node n ON (p.groupid = n.groupid AND n.noderole = 'primary')
WHERE logicalrelid::regclass = '$TABLE_NAME'::regclass;
```

Meaning, I think it's just a matter of updating their queries, there's not much engineering involved in adapting their scripts to the new schema.

Do our customers currently expect that they can simply upgrade and keep running Citus, or are they used to spending some time migrating? If they're used to migrating, I hope that as part of the migration to 7.0 asking them to add this `JOIN` wouldn't receive a lot of pushback?

**Comment by lithp** [comment]
My objection with creating a `pg_dist_placement` and turning `pg_dist_shard_placement` into a view is that this is confusing for new users.

They'll see both:

```
   View "pg_catalog.pg_dist_shard_placement"
   Column    |  Type   | Modifiers 
-------------+---------+-----------
 shardid     | bigint  | 
 shardstate  | integer | 
 shardlength | bigint  | 
 nodename    | text    | 
 nodeport    | bigint  | 
 placementid | bigint  | 
```

and 

```
                              Table "pg_catalog.pg_dist_placement"
   Column    |  Type   |                                   Modifiers                                   
-------------+---------+-------------------------------------------------------------------------------
 shardid     | bigint  | not null
 shardstate  | integer | not null
 shardlength | bigint  | not null
 placementid | bigint  | not null default nextval('pg_dist_shard_placement_placementid_seq'::regclass)
 groupid     | integer | not null
Indexes:
    "pg_dist_shard_placement_placementid_index" UNIQUE, btree (placementid)
    "pg_dist_shard_placement_shardid_index" btree (shardid)
    "pg_dist_shard_placement_groupid_index" btree (groupid)
Triggers:
    dist_placement_cache_invalidate AFTER INSERT OR DELETE OR UPDATE ON pg_dist_shard_placement FOR EACH ROW EXECUTE PROCEDURE master_dist_placement_cache_invalidate()
```

and wonder what the difference is between the two. It expands the window for users to write applications which hit `pg_dist_shard_placement`, making it even harder to eventually drop.

**Comment by marcocitus** [comment]
Are there backwards-compatible alternatives that would be less confusing? e.g. `pg_dist_shard_group`? or (slightly crazy) adding placement information to `pg_dist_shard`?

Probably worth noting that `pg_catalog`, in which the `pg_dist_*` tables reside, contain a mixture of views and real tables. For example, `pg_tables` is a view, while the actual table metadata is in `pg_class` (and `pg_namespace`, `pg_tablespace). Hopefully users peeking into pg_catalog tables realise the difference.

After this change, having a view that gives the nodename and nodeport of the primary node in a group for each shard placement without having to perform a join seems useful enough that we might want to add it if it wasn't already there, so it's not something we would necessarily ever remove.

**Comment by samay-sharma** [comment]
Not sure if anything is blocked on input from my side, but I think the view solution is better than removing it altogether.

Not many users look at what catalog tables Citus provides to see what to do. Mostly they touch / query the catalog tables on our recommendation. So, I'm not that concerned about them seeing two things and getting confused.

Also, for your question @lithp : I think the actual process is pretty simple, it is just around communicating to all the users and ensuring its done. On Citus Cloud for eg., the upgrades will be one-click in some time and maybe people don't read all docs before upgrading. (Not ideal but happens anyway)

So, whether we decide to fade it out over time or not, I think having a view makes sense.

**Comment by lithp** [comment]
Thanks for the discussion, incorporated this into #1449 and future work will proceed over there.
